test/semantics/omp-loop-association.f90

```diff
@@ -0,0 +1,140 @@
+! Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+! OPTIONS: -fopenmp
+
+! Check the association between OpenMPLoopConstruct and DoConstruct
+
+  integer :: b = 128
+  integer :: c = 32
+  integer, parameter :: num = 16
+  N = 1024
+
+! Different DO loops
+
+  !$omp parallel
+  !$omp do
+  do 10 i=1, N
+     a = 3.14
+10   print *, a
+  !$omp end parallel
+
+  !$omp parallel do
+  DO CONCURRENT (i = 1:N)
+     a = 3.14
+  END DO
+
+  !$omp parallel do simd
+  outer: DO WHILE (c > 1)
+     inner: do while (b > 100)
+        a = 3.14
+        b = b - 1
+     enddo inner
+     c = c - 1
+  END DO outer
+
+  c = 16
+  !ERROR: DO loop after the PARALLEL DO directive must have loop control
+  !$omp parallel do
+  do
+     a = 3.14
+     c = c - 1
+     if (c < 1) exit
+  enddo
+
+! Loop association check
+
+  ! If an end do directive follows a do-construct in which several DO
+  ! statements share a DO termination statement, then a do directive
+  ! can only be specified for the outermost of these DO statements.
+  do 100 i=1, N
+     !$omp do
+     do 100 j=1, N
+        a = 3.14
+100     continue
+    !ERROR: The ENDDO must follow the DO loop associated with the loop construct
```

Comment:
This error could be interpreted as referring to a normal Fortran END DO, rather than a $OMP END DO. I think at minimum messages that refer to OpenMP directives should say "directive", like the one above at line 48.

You should consider including $OMP in messages before directive names, e.g. "The $OMP ENDDO directive must follow...". If the user always looks at the source line displayed with the message, this may be redundant. But you can't always expect users to read carefully and including $OMP in the message immediately tells them the problem is related to the directive.

Either way, we need to decide on a convention and apply it consistently.

Reply:
Thanks for pointing that out. I will add `directive` in this PR. I think the second option is better and I will do a separate PR for that. With the `"!$OMP"`, it is easy to distinguish OpenMP msg with Fortran language itself or other programming models.